sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala

```diff
@@ -583,6 +611,24 @@ case class MyParser(spark: SparkSession, delegate: ParserInterface) extends Pars
     delegate.parseQuery(sqlText)
 }
 
+case class ExplainPlanInfoTagRule() extends Rule[SparkPlan] {
```

Comment:
This is really a hacky way to provide extended EXPLAIN info. Can we add a interface like
```
trait ExtendedExplainGenerator {
  def generateExtendedInfo(plan: SparkPlan): String
}
```

Then the config provides the implementation class.

Reply:
Let me see if I understand this right - 
A session extension implementation can implement the trait above and set the implementation in a config. If such an implementation is provided, then explain plan can print it. We can replace the `enableExtensionInfo` flag with this new conf.
The contents and  formatting of the extended explain string etc, now become part of the session extension and so we can remove the changes in `QueryPlan`. The extension implementors will have to do more work, but likely will have greater flexibility.
Let me know if I misunderstood. Otherwise I'll go ahead with this change.

Reply:
Thank you for the review, @cloud-fan .

Reply:
Yes, the EXPLAIN command will call this plugin implementation and append the result to the command output as extended information. Then we don't need to add new APIs in `QueryPlan` and the plugin can also be more flexible as it sees the full plan tree.